File: shared/src/iosMain/kotlin/gr/questweaver/shared/FlowUtils.kt
package gr.questweaver.shared

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import kotlin.coroutines.cancellation.CancellationException

inline fun <T : Any> Flow<T>.subscribe(
    scope: CoroutineScope,
    crossinline onError: (Throwable) -> Unit = {},
    crossinline onNext: (T) -> Unit,
) {
    scope.launch {
        try {
            collect {
                onNext(it)
            }
        } catch (ce: CancellationException) {
            throw ce
        } catch (t: Throwable) {
            onError(t)
        }
    }
}

File: iosApp/iosApp/Common/FlowUtils.swift
import Shared

extension Kotlinx_coroutines_coreFlow {
    func subscribe<A: AnyObject>(
        scope: Kotlinx_coroutines_coreCoroutineScope,
        onError: @escaping (KotlinThrowable) -> Void,
        onNext: @escaping (A?) -> Void
    ) {
        FlowUtilsKt.subscribe(
            self,
            scope: scope,
            onError: onError,
            onNext: { value in onNext(value as? A) }
        )
    }
}
